### PR TITLE
Add no_sync context when using DDP/FSDP and gradient accumulation

### PR DESCRIPTION
**Pull request by daniellepintz**
Summary:
if using gradient accumulation and DDP/FSDP, when in a step where we will not update the weights, run forward and backward in no_sync context
https://pytorch.org/docs/stable/_modules/torch/nn/parallel/distributed.html#DistributedDataParallel.no_sync
https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.no_sync

Reviewed By: ananthsub

Differential Revision: D40231819

### CODE REVIEW

**Review by ananthsub — APPROVED** [review]

**Comment by ananthsub** [comment]
merged here: https://github.com/pytorch/tnt/commit/a52ae6b57cf90d158eb8525cf7f268fefaaa5eba
